### PR TITLE
Migrate to Typescript

### LINKED ISSUE #66: Swap out Travis CI with GitHub Actions
Travis CI is no longer free for open source projects, so we need to move this project to GitHub Actions before any additional releases are made.

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
I have attempted to follow the lonlat transition "style" as closely as possible. Package's tests pass as do OTP-UI's when using this package.

Closes #66 

I'm just noticing now the function definition style doesn't match lonlat, please let me know if we'd like to change this.

@landonreed @evansiroky

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,14 +29,18 @@
   },
   "homepage": "https://github.com/conveyal/isomorphic-mapzen-search",
   "dependencies": {
-    "@conveyal/lonlat": "^1.1.2",
-    "isomorphic-fetch": "^2.2.1",
-    "qs": "^6.3.0"
+    "@conveyal/lonlat": "^1.4.1",
+    "isomorphic-fetch": "^3.0.0",
+    "qs": "^6.3.0",
+    "tsdx": "https://github.com/ibi-group/tsdx"
```
Comment (evansiroky) [inline package.json]: This should be a devDependency.

File: index.ts [hunk 2]
```diff
@@ -12,6 +11,48 @@ const autocompleteUrl = `${mapzenUrl}/autocomplete`
...
+type Query = {
+  // Can be disabled as it is a hack to conform with api
+  // eslint-disable-next-line camelcase
+  api_key: string
+  apiKey?: string
+  boundary?: Boundary
+  focusPoint?: LonLatInput
+  format?: boolean
+  layers?: string
+  options?: RequestInit
```
Comment (evansiroky) [inline index.ts]: I'm guessing we'll see this a lot more as we continue adding typings, but perhaps it'd be helpful to note that this is a built-in typing.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Looks great! Just one blocking comment about making tsdx a devDependency.

File: index.ts (hunk 2 above)
Comment (miles-grant-ibigroup) [reply index.ts]: I added a note, but I may have misunderstood what you meant!

**Comment by miles-grant-ibigroup** [comment]
Thank you! Have addressed comments, but may have misunderstood one of them

**Review by evansiroky — APPROVED** [review]

File: index.ts [hunk 3]
```diff
@@ -30,7 +70,7 @@ const searchUrl = `${mapzenUrl}/search`
  * @param {string} [$0.url='https://search.mapzen.com/v1/autocomplete']                       optional URL to override Mapzen autocomplete endpoint
  * @return {Promise}                              A Promise that'll get resolved with the autocomplete result
  */
-export function autocomplete ({
+export const autocomplete = ({
```
Comment (landonreed) [inline index.ts]: Generally I do prefer the `export function autocomplete (...) { }` syntax.

**Comment by miles-grant-ibigroup** [comment]
Thanks @landonreed, have changed the function syntax.

File: index.ts (hunk 3 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Changed in 9cef1b0e792b09793659b2070713873414974c77!

File: index.ts [hunk 4]
```diff
@@ -112,10 +149,10 @@ export function search ({
   sources = 'gn,oa,osm,wof',
   text,
   url = searchUrl
-}) {
+}: Query): Promise<Array<JSON>> {
```
Comment (landonreed) [inline index.ts]: Can we store `Promise<Array<JSON>>` in its own type? Looks to be reused a few times.

File: index.ts [hunk 5]
```diff
@@ -138,14 +175,17 @@ export function search ({
       query['boundary.rect.max_lon'] = boundary.rect.maxLon
     }
     if (boundary.circle) {
-      const {lat, lon} = lonlat(boundary.circle.centerPoint)
+      // TODO: not sure what to do here... It's clearly wrong
```
Comment (landonreed) [inline index.ts]: I'm not sure what this comment means. We should either fix (if it's a big concern) or add more context to the comment.

File: index.ts [hunk 6]
```diff
@@ -12,6 +11,47 @@ const autocompleteUrl = `${mapzenUrl}/autocomplete`
...
+  layers?: string
+  options?: RequestInit // Built-in Typing
+  point?: GeoJSON.Point | LonLatOutput
+  size?: number
+  sources?: string
+  text?: string
+  url?: string
+}
+
+type MapzenInput = {
```
Comment (landonreed) [inline index.ts]: This name feels weird to me. Maybe `PeliasFetchArgs` instead? (Mapzen no longer exists, even though it's still in the name of the repo.)

**Review by landonreed — APPROVED** [review]
Looks good! Just a couple of nits.

File: index.ts (hunk 5 above)
Comment (miles-grant-ibigroup) [reply index.ts]: I may have confused myself, but I think https://github.com/conveyal/lonlat/blob/948f87a1da22f3e4962841bd0c6b3e2e75e9a2d9/index.ts#L191 won't handle this single `centerPoint` correctly. But if the library has worked like this for years then maybe it doesn't make sense to change it? I've added a comment in 165129c2154d6b79a50f8b25ee3295a757fccfee, but I might remove it if I've made a mistake here...

File: index.ts (hunk 4 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Yes we can! In 165129c2154d6b79a50f8b25ee3295a757fccfee

File: index.ts (hunk 6 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Good call! Renamed in 165129c2154d6b79a50f8b25ee3295a757fccfee

**Comment by miles-grant-ibigroup** [comment]
Thanks for the feedback @landonreed!

File: index.ts (hunk 5 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Removed in c8987d80f93b051d137e728881e9eb94da1aab5b

**Pull request merged**